firmware-support/bittide-sys/src/gather_unit.rs

```diff
@@ -1,82 +1,33 @@
+impl<const MEM_SIZE: usize> GatherUnit<MEM_SIZE> {
+    /// Create a new [`GatherUnit`] instance given a base address.
     ///
     /// # Safety
     ///
-    /// The FDT must be a valid description of the hardware components that this
-    /// code is running on.
-    pub unsafe fn from_fdt_node(node: &FdtNode) -> Result<Self, ComponentLoadError> {
-        let get_node = |path| {
-            node.children()
-                .find(|child| matches_fdt_name(child, path))
-                .ok_or(ComponentLoadError::FdtNodeNotFound(path))
-        };
-
-        let get_reg = |node: &fdt::node::FdtNode, component| {
-            node.reg()
-                .ok_or(ComponentLoadError::RegNotFound { component })?
-                .next()
-                .ok_or(ComponentLoadError::RegNotFound { component })
-        };
-
-        let memory_node = get_node("gather-memory")?;
-        let metacycle_register_node = get_node("metacycle-reg")?;
-
-        let memory = get_reg(&memory_node, "memory")?;
-
-        if let Some(size) = memory.size {
-            if size != FRAME_SIZE {
-                return Err(ComponentLoadError::SizeMismatch {
-                    property: "memory frame size",
-                    expected: FRAME_SIZE,
-                    found: size,
-                });
-            }
+    /// The `base_addr` pointer MUST be a valid pointer that is backed
+    /// by a memory mapped gather unit instance.
+    pub unsafe fn new(base_addr: *const ()) -> GatherUnit<MEM_SIZE> {
+        let addr = base_addr as *const u8;
+        GatherUnit {
+            memory: addr.cast_mut(),
+            metacycle_register: addr.add(MEM_SIZE * 4),
         }
-
-        let metacycle_register = {
-            let reg = get_reg(&metacycle_register_node, "metacycle_register")?;
-            if let Some(size) = reg.size {
-                if size != 1 {
-                    return Err(ComponentLoadError::SizeMismatch {
-                        property: "metacycle register size",
-                        expected: 1,
-                        found: size,
-                    });
-                }
-            }
-            reg.starting_address
-        };
-
-        Ok(GatherUnit {
-            memory: memory.starting_address as *mut u8,
-            metacycle_register,
-        })
-    }
-
-    pub const fn frame_size(&self) -> usize {
-        FRAME_SIZE
     }
 
-    pub fn write_frame_memory(&mut self, data: &[u8; FRAME_SIZE]) {
-        for (i, d) in data.iter().copied().enumerate() {
-            unsafe {
-                self.memory.add(i).write_volatile(d);
-            }
-        }
+    /// # Safety
+    ///
+    /// The source memory size must be smaller or equal to the size of the
+    /// `GatherUnit` memory.
+    pub unsafe fn copy_from_slice(&self, src: &[u8], offset: usize) {
+        assert!(src.len() + offset <= MEM_SIZE * 4);
+        core::ptr::copy_nonoverlapping(src.as_ptr(), self.memory.add(offset), src.len());
```

Comment:
Can be removed if `Deref` and `DerefMut` are implemented.

Reply:
So this shouldn't be removed, but we can present a safe interface if you write the necessary asserts. If the `assert!` that already exists in this function is enough to guarantee that no out-of-bounds writes can occur, then it would be valid here to move the `unsafe` from the function to a block inside it. Also here, since this is MMIO, we mostly want to be using volatile operations to ensure the compiler doesn't elide them. What the HAL had previously is close to what it seems like you want, except for the ability to copy something smaller than a frame and have an offset. Something like
```rs
pub fn copy_from_slice(&self, src: &[u8], offset: usize) {
    assert!(src.len() + offset <= MEM_SIZE * 4);
    let mut off = offset;
    for &byte in src {
        unsafe {
            self.base_addr.add(off).write_volatile(byte);
        }
        off += 1;
    }
}
```
should probably get the job done.

Reply:
I've implemented your change request, but sadly this does make reading/writing roughtly 4 times slower than it was with `copy_nonoverlapping`. I've increased the padding of the calendar for the test by a factor of 4.
Do you think this change is worth it?
Do you agree with the new function names of the read/write functions for the scatter and gather units?